ractor/src/thread_local.rs

```diff
@@ -0,0 +1,305 @@
+// Copyright (c) Sean Lawlor
+//
+// This source code is licensed under both the MIT license found in the
+// LICENSE-MIT file in the root directory of this source tree.
+
+//! The thread-local module provides support for managing tasks which are not [Send]
+//! safe, and must remain pinned to the same thread for their lifetime.
+//!
+//! IMPORTANT: This ONLY works currently with Tokio's `rt` feature, specifically due to the usage
+//! of [tokio::task::LocalSet]
+
+use std::future::Future;
+
+use crate::concurrency::JoinHandle;
+use crate::ActorCell;
+use crate::ActorName;
+use crate::ActorProcessingErr;
+use crate::ActorRef;
+use crate::Message;
+use crate::RpcReplyPort;
+use crate::SpawnErr;
+use crate::State;
+use crate::SupervisionEvent;
+
+mod inner;
+
+/// Represents the state of an thread-local actor. Separate
+/// from the traditional [State] trait, this does NOT require
+/// that the state be [Send] and therefore does NOT need to be
+/// safe to send between threads.
+pub trait ThreadLocalState: std::any::Any + 'static {}
+impl<T: std::any::Any + 'static> ThreadLocalState for T {}
+
+/// [ThreadLocalActor] defines the behavior of an Actor. It specifies the
+/// Message type, State type, and all processing logic for the actor
+///
+/// NOTE: All of the implemented trait functions
+///
+/// * `pre_start`
+/// * `post_start`
+/// * `post_stop`
+/// * `handle`
+/// * `handle_serialized` (Available with `cluster` feature only)
+/// * `handle_supervisor_evt`
+///
+/// return a [Result<_, ActorProcessingError>] where the error type is an
+/// alias of [Box<dyn std::error::Error + Send + Sync + 'static>]. This is treated
+/// as an "unhandled" error and will terminate the actor + execute necessary supervision
+/// patterns. Panics are also captured from the inner functions and wrapped into an Error
+/// type, however should an [Err(_)] result from any of these functions the **actor will
+/// terminate** and cleanup.
+pub trait ThreadLocalActor: Default + Sized + 'static {
+    /// The message type for this actor
+    type Msg: Message;
+
+    /// The type of state this actor manages internally
+    type State: ThreadLocalState;
+
+    /// Initialization arguments
+    type Arguments: State;
+
+    /// Invoked when an actor is being started by the system.
```

Comment:
Maybe we can add a note here to explain that the inputs need to be `Send`, and only the resulting `State` will be `~Send`?